[Merged by Bors] - Allow per validator fee recipient via flag or file in validator client (similar to graffiti / graffiti-file)

pk910 opened the pull request:
## Issue Addressed

#2883 

## Proposed Changes

* Added `suggested-fee-recipient` & `suggested-fee-recipient-file` flags to validator client (similar to graffiti / graffiti-file implementation).
* Added proposer preparation service to VC, which sends the fee-recipient of all known validators to the BN via [/eth/v1/validator/prepare_beacon_proposer](https://github.com/ethereum/beacon-APIs/pull/178) api once per slot
* Added [/eth/v1/validator/prepare_beacon_proposer](https://github.com/ethereum/beacon-APIs/pull/178) api endpoint and preparation data caching
* Added cleanup routine to remove cached proposer preparations when not updated for 2 epochs

## Additional Info

Changed the Implementation following the discussion in #2883.

pk910 committed:
implemented per validator fee-recipient (fee-recipient / fee-recipient-file flags in validator client)

pk910 committed:
run cargo fmt

pk910 committed:
fix clippy issues

pk910 commented:
Hmm, it was a nice try, but doesn't work at the moment...
> lighthouse[69268]: Jan 17 09:34:48.001 CRIT Error whilst producing block            message: All endpoints failed http://localhost:5052/ => RequestFailed("Error from beacon node when producing block: ServerMessage(ErrorMessage { code: 400, message: \"BAD_REQUEST: invalid query: Invalid query string\", stacktraces: [] })"), service: block

Seems I missed something - Looking into it..

pk910 committed:
fixed fee_recipient parameter formatting

pawanjay176 commented:
@pk910 This is looking good :) Your issue looks like incorrect parsing of the address. You can try printing out the `fee_recipient` right before it is passed to the api in `block_service.rs` and see if it's what you are expecting. 

> Is it allowed to add an optional parameter to the bn api just like that or is the api specification managed somewhere else and needs a change request there?

You would need to request for change in the api specification as having a different behaviour would break lighthouse vc compatibility with other client beacon nodes. The api specification is maintained here https://github.com/ethereum/beacon-apis/issues

pk910 commented:
found the problem. VC was passing the fee_recipient with a trimmed formatting (0x0000…0001).
testing it now and will mark as ready for review again once I've successfully proposed a block in kintsugi.

pk910 committed:
reverted fee_recipient api parameter, added proposer preparation service that generates proposer preparations for all known validators at the beginning of each epoch

pk910 committed:
Added prepare_beacon_proposer api endpoint

pk910 committed:
cache proposer preparations

pk910 committed:
Fixed VC panic when fee-recipient-file is unreadable & reload cached file contents

pk910 committed:
Merge remote-tracking branch 'origin/unstable' into fee-recipient-per-validator

# Conflicts:
#	consensus/types/src/lib.rs

pk910 committed:
cleanup implementation

pk910 commented:
I've reimplemented the way how the fee-recipient is transmitted to the BN. There is now a new "proposal preparation" service in VC that sends the fee-recipients of all known validators to the BN at the beginning of each epoch using the new /eth/v1/validator/prepare_beacon_proposer api.

This is a working prototype that has successfully proposed blocks with individual fee-recipients for me in the kintsugi net.

A few things that should be looked into:
* VC publishes the proposer preparations at the beginning of each epoch. I'm not sure if that interferes other much more important api calls at the beginning of an epoch or could overload the BN with a big number of VCs?
* The proposer preparation is currently stored forever in a simple HashMap.. I'm pretty sure there is a much more elegant solution for that and I have a feeling that this cache should not live in the execution layer code (actually I didn't get it running in the BeaconChain struct, which looks like a better place for such a cache)
* Proposer preparations are currently stored in memory forever. There should be something that removes them again when not updated for 2 epochs. As I think the cache implementation needs rework anyway I have not invested work in that yet.
* There is no limitation of how many proposer preparations are stored in memory.. As the index is a u64 there can probably be enough entries to fill up BN memory completely - that looks like a possible dos attack surface to me

It can probably be implemented in a much cleaner way, so I'm totally fine if you decide to throw it away and redo yourself ;)
Can also try to rework some parts myself if you give me some advice how to continue :)

paulhauner commented:
I haven't done a review, but thanks for your contribution @pk910!

Earlier today I created https://github.com/sigp/lighthouse/issues/2936. It seems this PR is along similar lines :tada: 

I'll find some time this week or early next to review this :pray:

paulhauner committed:
Add suggested fee recip. to val defs yaml

paulhauner committed:
Use fee recip from val defs

paulhauner committed:
Add incomplete changes

paulhauner committed:
Add incomplete docs

paulhauner committed:
Tidy, publish each slot

paulhauner committed:
Add "suggested-" prefix globally

paulhauner committed:
Add comment

paulhauner committed:
Tidy docs

paulhauner committed:
Clarify docs

paulhauner committed:
Tidy logs

paulhauner committed:
Publish on startup

paulhauner committed:
Add val defs tests

paulhauner committed:
Fix test compilation failures

paulhauner committed:
Fix web3signer test failures

paulhauner committed:
Fix account manager tests

paulhauner reviewed (changes requested):
This is awesome, thank you! It's consistent with our existing code and very maintainable :pray: 

I've made some suggestions and added some features over in https://github.com/pk910/lighthouse/pull/1.

Let me what you think about my suggestions, I'm looking forward to seeing this merged :relaxed:

pk910 committed:
Merge pull request #1 from paulhauner/2924-paul

Suggestions for fee recipieint changes from paulhauner

pk910 commented:
Thanks for your review & additions, I've merged your changes in here.
That makes the fee-recipient fully configurable through the validator store, too 🎉

I'm a little bit worried about the removal of the hardcoded default fee-recipient.
What happens if no fee-recipient is specified on VC side and no `--suggested-fee-recipient` flag is given on BN side?
I think that would currently make the BN throwing a [`Error::FeeRecipientUnspecified`](https://github.com/pk910/lighthouse/blob/87a0e14adc6ad54b75e5308b59e73f584b3e87d2/beacon_node/execution_layer/src/lib.rs#L294) error on [`ExecutionLayer::get_payload`](https://github.com/pk910/lighthouse/blob/87a0e14adc6ad54b75e5308b59e73f584b3e87d2/beacon_node/execution_layer/src/lib.rs#L314) call, that prevents the BN from building new EL blocks?
If so I'd leave the hardcoded default with an unusable address like "0x0000000000000000000000000000000000000000".
Even if there is no financial benefit for the proposer, its better for the network to produce a block without getting any fees than not producing a EL block at all?
Especially for the merge transition where probably not every staking user has seen this new fee-recipient setting and is running his/her VC & BN without a fee-recipient.

No suggestions regarding the ProposerPreparationData caching in a simple HashMap in the ExecutionLayer crate?
If that is ok for you there is at least one thing missing:
There should be a cleanup routine that removes old ProposerPreparationData if not updated for some time (proposal says 2 epochs). Otherwise these entries are stored in memory forever...
I can try on implementing this - just didn't want to start that yet as I thought this cache will be reworked anyway ☺️

pk910 commented:
Should we publish these proposer preparations to all BN's or just the [first fully synced one](https://github.com/pk910/lighthouse/blob/87a0e14adc6ad54b75e5308b59e73f584b3e87d2/validator_client/src/preparation_service.rs#L261-L267) as it currently does?
I think it would be better to publish it to all available BNs as its just preparation data and does not do anything immediatly..
On the other side that would cause recurring calls to fallback BNs, which is not ideal when using a 3rd-party fallback BN like infura where you have a limited contingent of calls / traffic.

And why do we need to publish the proposal preparations each slot instead of each epoch?

pk910 committed:
Added client IP to "Received proposer preparation data" log entry

pk910 committed:
Added optional 'suggested_fee_recipient' field in 'POST /lighthouse/validators/' & 'POST /lighthouse/validators/web3signer' documentation

paulhauner commented on validator_client/src/preparation_service.rs:
```diff
@@ -0,0 +1,283 @@
+
+    /// Prepare proposer preparations and send to beacon node
+    async fn prepare_proposers_and_publish(&self, spec: &ChainSpec) -> Result<(), String> {
+        let preparation_data = self.collect_preparation_data(spec).unwrap();
```
We need to handle this `Result` without a runtime panic. I think `collect_preparation_data` doesn't need to return a `Result`, so it's an easy fix.

paulhauner commented on beacon_node/execution_layer/src/lib.rs:
```diff
@@ -239,6 +243,51 @@ impl ExecutionLayer {
+        self.block_on_generic(|_| async move {
+            self.update_proposer_preparation(preparation_data).await
+        })
+        .unwrap()
```
The general ethos (with exceptions) in Lighthouse is "never panic", so we need to gracefully handle this `Result`, otherwise this function might panic during shutdown.

pk910 replied:
Thanks for reviewing properly 👍
Thats the result of not understanding what is going on completely, yet. 😅
I think I got it now and will recheck the rest of my changes for similar issues 😊

For here I think it's best to propagate the error up to the calling function via `?`.
The [calling function](https://github.com/pk910/lighthouse/blob/96d5cf9c0b75d603e7f28f0c9eb127be7ae5b749/beacon_node/http_api/src/lib.rs#L2209-L2215) already handles errors properly and rejects the api call with a `custom_bad_request` in case of an error.
Is that correct?

paulhauner commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -2175,6 +2175,48 @@ pub fn serve<T: BeaconChainTypes>(
+                        log,
+                        "Received proposer preparation data";
+                        "count" => preparation_data.len(),
+                        "client" => format!("{:?}", client_addr.unwrap()),
```
```suggestion
                        "client" => client_addr
                            .map(|a| a.to_string())
                            .unwrap_or_else(|| "unknown".to_string()),
```

Otherwise this function will panic if `client_addr` is `None`.

paulhauner commented on beacon_node/execution_layer/src/lib.rs:
```diff
@@ -254,8 +303,10 @@ impl ExecutionLayer {
         timestamp: u64,
         random: Hash256,
         finalized_block_hash: Hash256,
+        proposer_index: Option<u64>,
```
I think we can just use a `u64` here, rather than an `Option<u64>`. It would make the function less flexible, but I think the simplicity is worth it.

paulhauner replied:
I ended up addressing this in https://github.com/paulhauner/lighthouse/commit/07a6a43eaf06d59334177c82293b73ce9f7e95a8.

See https://github.com/sigp/lighthouse/pull/2924#issuecomment-1025376029

pk910 replied:
Used an `Option<u64>` there to be able to pass `None` in the [test code](https://github.com/sigp/lighthouse/blob/07a6a43eaf06d59334177c82293b73ce9f7e95a8/beacon_node/execution_layer/src/test_utils/mock_execution_layer.rs#L130-L139) that used that function 😉
But just passing 0 as `proposer_index` in the test seems fine, too and makes the Option unnecessary 👍

paulhauner committed:
Add default fee recip to BN

paulhauner commented:
> If so I'd leave the hardcoded default with an unusable address like "0x0000000000000000000000000000000000000000".
Even if there is no financial benefit for the proposer, its better for the network to produce a block without getting any fees than not producing a EL block at all?

Hmm, yeah this is a tricky one. It is better for the network to have blocks, but also feels a bit wrong to make assumptions about the fee address for users.

Something I noticed is that the `prepare_beacon_proposers` endpoint has a race-condition where VCs might want to produce blocks before they've told the BN what their fee recip. should be. I've suggested a fix here: https://github.com/ethereum/beacon-APIs/issues/187

Especially given this quirk in the API, I think that you're right and it's sensible to leave a default value. I've fixed this again in https://github.com/paulhauner/lighthouse/commit/07a6a43eaf06d59334177c82293b73ce9f7e95a8, it should be easy to cherry-pick that in :pray: 

> No suggestions regarding the ProposerPreparationData caching in a simple HashMap in the ExecutionLayer crate?

I think the `HashMap` is great! Simple and effective.

You may want to consider a `RwLock` instead of a `Mutex`, so you can have concurrent readers. However, I'm not sure that's of huge real-world gain.

> There should be a cleanup routine that removes old ProposerPreparationData if not updated for some time (proposal says 2 epochs). Otherwise these entries are stored in memory forever...
> I can try on implementing this - just didn't want to start that yet as I thought this cache will be reworked anyway

If you're happy to implement this, I think it would be great! :pray: 

> Should we publish these proposer preparations to all BN's or just the first fully synced one as it currently does?

I think we should stick with the status-quo of just publishing to the primary one for now. I'm keen to one-day implement something more complex, but I think we should do that separately.

> And why do we need to publish the proposal preparations each slot instead of each epoch?

There's no hard rule here, I just thought per-slot is a little safer and the impact is minor. Sorry I didn't explain my intentions earlier.

pk910 committed:
Update beacon_node/http_api/src/lib.rs

Co-authored-by: Paul Hauner <paul@paulhauner.com>

pk910 committed:
Merge remote-tracking branch 'origin/unstable' into fee-recipient-per-validator

pk910 committed:
fix merge issue

pk910 committed:
fixed result handling in validator_client/src/preparation_service.rs

pk910 committed:
fixed result handling in beacon_node/execution_layer/src/lib.rs

pk910 committed:
cargo fmt

pk910 committed:
fixed another merge from unstable issue

pk910 committed:
implemented preparation data cleanup routine

pk910 commented:
@paulhauner 
I've added your suggestions and implemented the cache cleanup routine.
Tested locally with more debug output - it removes proposer preparations when not updated for 2 epochs now.
Ready for hopefully final review 😊

pk910 committed:
dummy commit to trigger test-suite again

pk910 committed:
shorten "Successfully published proposer preparation" log entry (it's too long and looks bad in log file :D)

paulhauner commented on beacon_node/beacon_chain/src/execution_payload.rs:
```diff
@@ -273,6 +273,7 @@ pub async fn prepare_execution_payload<T: BeaconChainTypes>(
     let timestamp = compute_timestamp_at_slot(state, spec).map_err(BeaconStateError::from)?;
     let random = *state.get_randao_mix(state.current_epoch())?;
     let finalized_root = state.finalized_checkpoint().root;
+    let proposer_index = state.get_beacon_proposer_index(state.slot(), spec)? as u64;
```
I think it would be best if we pass `proposer_index` as a value to `prepare_execution_payload`. The `BeaconState::get_beacon_proposer_index` is somewhat heavy (it involves crypto) and we've seen it impact performance. The proposer index is known when producing a block, so it should be super easy to thread through :pray:

paulhauner commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -2186,6 +2186,53 @@ pub fn serve<T: BeaconChainTypes>(
+                        .epoch()
+                        .map_err(warp_utils::reject::beacon_chain_error)?;
+
+                    info!(
```
```suggestion
                    debug!(
```

This might get a little noisy when polling once per slot (per validator client).

pk910 replied:
Agree 👍
Should we change the [log entry on VC side](https://github.com/pk910/lighthouse/blob/3072e5a25938a6f01003179928d9bbd72c27be04/validator_client/src/preparation_service.rs#L265-L269) to `debug`, too?

paulhauner replied:
Oh yep, good idea!

paulhauner commented on validator_client/src/config.rs:
```diff
@@ -41,6 +42,11 @@ pub struct Config {
     /// Graffiti file to load per validator graffitis.
     pub graffiti_file: Option<GraffitiFile>,
+    /// fee_recipient address, that will receive transaction fees collected from any blocks
+    /// produced by this node.
```
```suggestion
    /// Fallback fallback address.
```

Just so it doesn't sound like the canonical source.

paulhauner commented on validator_client/src/cli.rs:
```diff
@@ -127,6 +127,21 @@ pub fn cli_app<'a, 'b>() -> App<'a, 'b> {
+            Arg::with_name("suggested-fee-recipient")
+                .long("suggested-fee-recipient")
+                .help("Specify the address, that will receive transaction fees collected from any \
+                        blocks produced by this validator client.")
```
```suggestion
                .help("The fallback address provided to the BN if nothing suitable is found \
                           in the validator definitions or fee recipient file.")
```

It's a messy diff sorry, but I'm trying to express the ordering.

paulhauner commented on validator_client/src/cli.rs:
```diff
@@ -127,6 +127,21 @@ pub fn cli_app<'a, 'b>() -> App<'a, 'b> {
+        .arg(
+            Arg::with_name("suggested-fee-recipient-file")
+                .long("suggested-fee-recipient-file")
+                .help("Specify a fee-recipient file to load validator fee-recipients from.")
```
```suggestion
                .help("The fallback address provided to the BN if nothing suitable is found \
                           in the validator definitions.")
```

paulhauner reviewed (changes requested):
This is looking really good, we're *so* close. I've left some minor requests here, I'm keen to merge after they're addressed.

Thanks again for this high-value contribution :pray:

pk910 committed:
Apply suggestions from code review

Co-authored-by: Paul Hauner <paul@paulhauner.com>

pk910 committed:
pass proposer_index as value to prepare_execution_payload

pk910 commented:
@paulhauner 
All suggestions applied 👍

paulhauner commented on validator_client/src/preparation_service.rs:
```diff
@@ -0,0 +1,278 @@
+            })
+            .await
+        {
+            Ok(()) => info!(
```
```suggestion
            Ok(()) => debug!(
```

paulhauner commented on validator_client/src/preparation_service.rs:
```diff
@@ -0,0 +1,278 @@
+    validator_store::{DoppelgangerStatus, ValidatorStore},
+};
+use environment::RuntimeContext;
+use slog::{error, info};
```
```suggestion
use slog::{debug, error, info};
```

paulhauner reviewed (changes requested):
This looks great, I just made suggestions for your comment in https://github.com/sigp/lighthouse/pull/2924#discussion_r800809264.

After those a resolved, it's merge time :tada:

pk910 committed:
Apply suggestions from code review

Co-authored-by: Paul Hauner <paul@paulhauner.com>

pk910 commented:
@paulhauner 
applied 👍
ready to merge? 🎉

paulhauner commented:
Let's goooo

bors r+

paulhauner commented:
bors retry
